dali/pipeline/operators/reader/loader/video_loader.h

```diff
@@ -0,0 +1,219 @@
+     *  than packets_decoded because decoding must happen key frame to
+     *  key frame and output sequences often span key frame sequences,
+     *  requiring more frames to be decoded than are actually used in
+     *  the output. */
+    uint64_t frames_used;
+};
+
+
+class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
+ public:
+  explicit inline VideoLoader(const OpSpec& spec,
+    const std::vector<std::string>& filenames)
+    : Loader<GPUBackend, SequenceWrapper>(spec),
+      count_(spec.GetArgument<int>("count")),
+      output_height_(spec.GetArgument<int>("height")),
+      output_width_(spec.GetArgument<int>("width")),
+      height_(0),
+      width_(0),
+      filenames_(filenames),
+      // TODO(spanev) handle device_id != 0
+      device_id_(0),
+      codec_id_(0),
+      done_(false) {
+  }
+
+  void init() {
+
+    /* Required to use libavformat: Initialize libavformat and register all
+     * the muxers, demuxers and protocols.
+     */
+    av_register_all();
+
+    // TODO(spanev) Implem several files handling
+    total_frame_count_ = get_or_open_file(filenames_[0]).frame_count_;
+
+    // TODO(spanev) to change after deciding what we want
+    int seq_per_epoch = total_frame_count_ / count_;
+    frame_starts_.resize(seq_per_epoch);
+    for (int i = 0; i < seq_per_epoch; ++i) {
+      frame_starts_[i] = i * count_;
+    }
+    // TODO(spanev) change random engine
+    auto rng = std::default_random_engine {};
+    std::shuffle(std::begin(frame_starts_), std::end(frame_starts_), rng);
+    current_frame_idx_ = 0;
+
+    thread_file_reader_ = std::thread{&VideoLoader::read_file, this};
+  }
+
+  ~VideoLoader() noexcept {
+    done_ = true;
+    send_queue_.cancel_pops();
+    if (vid_decoder_) {
+      vid_decoder_->finish();
+    }
+    if (thread_file_reader_.joinable()) {
+      try {
+        thread_file_reader_.join();
+      } catch (const std::system_error& e) {
+        DALI_FAIL("System error joining thread: "
+                   + e.what());
+      }
+    }
+  }
+
+  void PrepareEmpty(SequenceWrapper *tensor) override;
+  void ReadSample(SequenceWrapper *tensor) override;
+  Index Size() override;
+
+  OpenFile& get_or_open_file(std::string filename);
+  void seek(OpenFile& file, int frame);
+  void read_file();
+  void push_sequence_to_read(std::string filename, int frame, int count);
+  void receive_frames(SequenceWrapper& sequence);
+
+  // Params
+  int count_;
+  int output_height_;
+  int output_width_;
+  int height_;
+  int width_;
+  std::vector<std::string> filenames_;
+
+  int device_id_;
+  int codec_id_;
+  VideoLoaderStats stats_;
+
+  std::unordered_map<std::string, OpenFile> open_files_;
+  std::unique_ptr<NvDecoder> vid_decoder_;
+
+  ThreadSafeQueue<FrameReq> send_queue_;
+
+  std::thread thread_file_reader_;
+
+  int total_frame_count_;
+  std::vector<int> frame_starts_;
+  unsigned current_frame_idx_;
+
+  bool done_;
```

Comment:
Shouldn't that be a volatile variable?

Reply:
Do you mean atomic? I can't get why this should be volatile

Reply:
Basically we should all variable that are used to communicate with thread or atomic or at least volatile.
I see that compiler can change:
```
while(run) {
}
```
to:
```
if (run) {
  while (true) {
  }
}
```
@mzient - any thoughts?

Reply:
Ok I was not aware of that, I though that this optimization could happen only if the variable is not modifier anywhere else in the scope (the scope here being the class)

Reply:
No, for compiler the scope is the function. And even if it it just a global variable this need to be `volatile` if you want to use it for inter thread communication.

Reply:
Done